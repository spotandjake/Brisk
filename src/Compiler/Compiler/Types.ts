import { Position } from '../Types';
import { Stack } from './Helpers'; //TODO: replace helpers with normal stuff
// Lexer
export const enum LexemeType {
  keyword,
  separator,
  operator,
  literal,
  flag,
  comment,
  identifier
}
interface Lexeme {
  type: LexemeType;
  id: string;
  value?: (text: string) => string | number | boolean | bigint;
  match: RegExp;
  lineBreaks?: boolean;
}
interface Token {
  type: string;
  value: string | number | boolean | bigint;
  text: string;
  offset: number;
  line: number;
  col: number;
  file: string;
}
// Parser
export const enum HeapTypeID {
  None = 0,
  Function = 1,
  Closure = 2,
  Boolean = 3, // Deprecated
  String = 4,
  Number = 5,
  Array = 6,
  Parameters = 7
}
export const enum ParseTreeNodeType {
  Program,
  importStatement,
  importWasmStatement,
  exportStatement,
  declarationStatement,
  callStatement,
  flagStatement,
  commentStatement,
  blockStatement,
  ifStatement,
  variable,
  literal,
  functionNode,
  functionParameter,
  functionType
}
export type ParseTreeNode = Program | Statement | ExpressionNode | FunctionParameterNode;
export type Import = (boolean | string[]);
export interface Program {
  type: ParseTreeNodeType.Program;
  flags: FlagStatementNode[];
  variables?: Stack;
  globals?: any[];
  body: Statement[];
  exports: string[];
  imports: { path: string; identifiers: Import; }[];
  position: Position;
}
export type Statement =
  ImportStatementNode | ImportWasmStatementNode | ExportStatementNode |
  DeclarationStatementNode | CallStatementNode | FlagStatementNode |
  CommentStatementNode | BlockStatementNode | IfStatementNode;
export interface ImportStatementNode {
  type: ParseTreeNodeType.importStatement;
  identifier: string;
  path: string;
  position: Position;
}
export interface ImportWasmStatementNode {
  type: ParseTreeNodeType.importWasmStatement;
  dataType: TypeNode;
  identifier: string;
  path: string;
  position: Position;
}
export interface ExportStatementNode {
  type: ParseTreeNodeType.exportStatement;
  identifier: string;
  position: Position;
}
export interface DeclarationStatementNode {
  type: ParseTreeNodeType.declarationStatement;
  dataType: TypeNode;
  identifier: string;
  value: ExpressionNode;
  position: Position;
}
export interface CallStatementNode {
  type: ParseTreeNodeType.callStatement;
  identifier: string;
  arguments: ExpressionNode[];
  position: Position;
}
export interface FlagStatementNode {
  type: ParseTreeNodeType.flagStatement;
  value: string;
  position: Position;
}
export interface CommentStatementNode {
  type: ParseTreeNodeType.commentStatement;
  value: string;
  position: Position;
}
export interface BlockStatementNode {
  type: ParseTreeNodeType.blockStatement;
  variables?: Stack;
  body: Statement[];
  position: Position;
}
export interface IfStatementNode {
  type: ParseTreeNodeType.ifStatement;
  condition: ExpressionNode;
  body: Statement;
  alternate?: Statement;
  position: Position;
}
export type ExpressionNode =
  LiteralNode | FunctionNode | CallStatementNode | VariableNode;

export interface VariableNode {
  type: ParseTreeNodeType.variable;
  identifier: string;
  position: Position;
}
export interface LiteralNode {
  type: ParseTreeNodeType.literal;
  dataType: TypeNode;
  value: string | number | boolean | bigint;
  position: Position;
}
export interface FunctionNode {
  type: ParseTreeNodeType.functionNode;
  dataType: TypeNode;
  flags: FlagStatementNode[];
  variables?: Stack;
  parameters: FunctionParameterNode[];
  body: Statement[];
  position: Position;
}
export interface FunctionParameterNode {
  type: ParseTreeNodeType.functionParameter;
  dataType: TypeNode;
  identifier: string;
  position: Position;
}
export type TypeNode = FunctionTypeNode | string;
export interface FuncTypeNode {
  value: FunctionTypeNode;
}
export interface FunctionTypeNode {
  type: ParseTreeNodeType.functionType;
  params: TypeNode[];
  result: TypeNode;
}
// Exports
export {
  // Lexer
  Lexeme,
  Token,
  // Parser
};