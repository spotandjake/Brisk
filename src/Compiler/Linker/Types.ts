import binaryen from 'binaryen';
export interface WasmFunction {
  name: string;
  params: binaryen.Type;
  results: binaryen.Type;
  vars: binaryen.Type[];
  body: binaryen.ExpressionRef;
}
export interface FunctionImport {
  type: 'FunctionImport';
  value: {
    internalName: string;
    externalModuleName: string;
    externalBaseName: string;
    params: binaryen.Type;
    results: binaryen.Type;
  }
}
export interface GlobalImport {
  type: 'GlobalImport';
  value: {
    internalName: string;
    externalModuleName: string;
    externalBaseName: string;
    globalType: binaryen.Type;
  }
}
export type PoolImport = FunctionImport | GlobalImport;
export const enum ModuleType {
  BriskModule,
  WasmModule,
  LocalModule
}
export interface Dependency {
  entry: boolean;
  importance: number;
  found: boolean;
  location: string;
  binaryen_module?: binaryen.Module;
}
export interface Pool {
  globals: Map<string, string>;
  functions: Map<string, string>;
  // Wasm Level Imports, Exports
  imports: Map<string, string>;
  exports: Map<string, string>; //TODO: this is gonna need more info
}
export interface MergePool {
  globals: { name: string; type: binaryen.Type; mutable: boolean; init: binaryen.ExpressionRef; }[];
  functions: WasmFunction[];
  functionTable: string[];
  // Wasm Level Imports, Exports
  imports: PoolImport[];
  exports: string[]; //TODO: this is gonna need more info
}
export interface CountPool { // Used For Keeping Track Of Number Of Parts
  globals: number;
  functions: number;
}