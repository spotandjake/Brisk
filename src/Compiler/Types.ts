interface Position {
  offset: number;
  line: number;
  col: number;
  file: string;
}
interface CompilerOptions {
  wat?: boolean;
  link?: boolean;
  writeFile?: boolean;
}
export {
  Position,
  CompilerOptions
};