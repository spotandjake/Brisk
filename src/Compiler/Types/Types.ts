interface Position {
  offset: number;
  line: number;
  col: number;
  file: string;
}
// Exports
export {
  Position
};