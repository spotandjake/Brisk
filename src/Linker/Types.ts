// Types
export interface BriskModule {
  moduleTypeSignature: {
    Types: [];
    Exports: [];
  };
}
