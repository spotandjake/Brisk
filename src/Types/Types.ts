export interface Position {
  offset: number;
  length: number;
  line: number;
  col: number;
  file: string;
}
