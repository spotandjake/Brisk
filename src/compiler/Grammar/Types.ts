// Type Imports
import { Stack } from '../Helpers/Helpers';
import * as path from 'path';
// General Nodes
export interface Position {
  offset: number;
  line: number;
  col: number;

  file?: path.ParsedPath;
}
export interface Token {
  type: string;
  value: string | number | boolean | bigint;
  text: string;
  offset: number;
  line: number;
  col: number;
}
// ParseTreeTypes
export type ParseTreeNode = Program | ProgramNode | Statement | ExpressionNode | FunctionParameterNode;
export interface LinkedModule {
  type: 'Program';
  flags: FlagStatementNode[];
  variables: Stack;
  body: Statement[];
  exports: string[];
  imports: string[];
  imported: string[];
  position: Position;
}
export interface Program {
  type: 'Program';
  flags: FlagStatementNode[];
  variables: Stack;
  body: Statement[];
  exports: string[];
  imports: string[];
  position: Position;
}
export interface ProgramNode {
  type: 'Program';
  body: Statement[];
  exports?: string[];
  imports?: string[];
  position: Position;
}
export type Statement = 
  ImportStatementNode      | ImportWasmStatementNode | ExportStatementNode |
  DeclarationStatementNode | CallStatementNode       | FlagStatementNode   |
  CommentStatementNode     | BlockStatementNode;
export interface ImportStatementNode {
  type: 'importStatement';
  identifier: string;
  path: string;
  position: Position;
}
export interface ImportWasmStatementNode {
  type: 'importWasmStatement';
  dataType: TypeNode;
  identifier: string;
  path: string;
  position: Position;
}
export interface ExportStatementNode {
  type: 'exportStatement';
  identifier: string;
  position: Position;
}
export interface DeclarationStatementNode {
  type: 'declarationStatement';
  dataType: TypeNode;
  identifier: string;
  value: ExpressionNode;
  position: Position;
}
export interface CallStatementNode {
  type: 'callStatement';
  identifier: string;
  arguments: ExpressionNode[];
  position: Position;
}
export interface FlagStatementNode {
  type: 'flagStatement';
  value: string;
  position: Position;
}
export interface CommentStatementNode {
  type: 'commentStatement';
  value: string;
  position: Position;
}
export interface BlockStatementNode {
  type: 'blockStatement';
  variables?: Stack;
  body: Statement[];
  position: Position;
}
export type ExpressionNode = 
  LiteralNode | FunctionDeclarationNode | FunctionNode | CallStatementNode | VariableNode;

export interface VariableNode {
  type: 'variable';
  identifier: string;
  position: Position;
}
export interface LiteralNode {
  type: 'literal';
  dataType: TypeNode;
  value: string | number | boolean | bigint;
  position: Position;
}
export interface FunctionNode {
  type: 'functionNode';
  dataType: TypeNode;
  flags: FlagStatementNode[];
  variables: Stack;
  parameters: FunctionParameterNode[];
  body: Statement[];
  position: Position;
}
export interface FunctionDeclarationNode {
  type: 'functionDeclaration';
  dataType: TypeNode;
  parameters: FunctionParameterNode[];
  body: Statement[];
  position: Position;
}
export interface FunctionParameterNode {
  type: 'functionParameter';
  dataType: TypeNode;
  identifier: string;
  position: Position;
}
export type TypeNode = FunctionTypeNode | string;
export interface FuncTypeNode {
  value: FunctionTypeNode;
}
export interface FunctionTypeNode {
  type: 'functionType';
  params: TypeNode[];
  result: TypeNode;
}