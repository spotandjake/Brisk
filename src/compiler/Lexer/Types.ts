export interface Token {
  type: string;
  value: string | number | boolean | bigint;
  text: string;
  offset: number;
  line: number;
  col: number;
}
export interface Rule {
  type: string;
  id: string;
  value?: (text: string) => string | number | boolean | bigint;
  match: RegExp;
  lineBreaks?: boolean;
}