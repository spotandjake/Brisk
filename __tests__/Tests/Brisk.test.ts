// Test Utils
//@ts-ignore
import { test, expect } from '@jest/globals';
import fs from 'fs';
// Test Components
import lex from '../../src/Compiler/Lexer/index';
import parse from '../../src/Compiler/Parser/index';
import analyze from '../../src/Compiler/Analyzer/index';
import compile from '../../src/Compiler/index';
import { ExportList } from '../../src/Compiler/Types/Types';
// Import Data: Relative to dist
const Main_Pass = fs.readFileSync('./__tests__/Data/Brisk/Brisk_pass.br', 'utf8');
const Literals_Pass = fs.readFileSync('./__tests__/Data/Brisk/Literals_pass.br', 'utf8');
// PolyFill Compile
const compileFile = async (filePath: string): Promise<{ output: '', exports: ExportList }> => {
  // TODO: Return Proper Types
  return {
    output: '',
    exports: new Map()
  };
};
// Compiler
test('Compile: Main Pass', async () => {
  expect(await compile(Main_Pass, 'file', compileFile)).toMatchSnapshot();
});
// Parser-Lexer
test('Parser-Lexer: Main Pass', () => {
  expect(parse(lex(Main_Pass, 'file'), '', 'file')).toMatchSnapshot();
});
// Parser-Lexer-Analyzer
test('Parser-Lexer-Analyzer: Main Pass', () => {
  expect(analyze(Main_Pass, parse(lex(Main_Pass, 'file'), '', 'file'))).toMatchSnapshot();
});
// Literals
test('Compile: Literals Pass', async () => {
  expect(await compile(Literals_Pass, 'file', compileFile)).toMatchSnapshot();
});
// Parser-Lexer
test('Parser-Lexer: Literals Pass', () => {
  expect(parse(lex(Literals_Pass, 'file'), '', 'file')).toMatchSnapshot();
});
// Parser-Lexer-Analyzer
test('Parser-Lexer-Analyzer: Literals Pass', () => {
  expect(analyze(Literals_Pass, parse(lex(Literals_Pass, 'file'), '', 'file'))).toMatchSnapshot();
});
