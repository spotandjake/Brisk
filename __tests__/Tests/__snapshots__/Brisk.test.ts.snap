// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compile: Main 1`] = `
Object {
  "body": Array [],
  "category": 0,
  "nodeType": 0,
  "position": Object {
    "col": 0,
    "file": "file",
    "line": 0,
    "offset": 0,
  },
  "stack": Map {},
  "variables": Map {},
}
`;

exports[`Parser-Lexer: Main 1`] = `
Object {
  "body": Array [],
  "category": 0,
  "nodeType": 0,
  "position": Object {
    "col": 0,
    "file": "file",
    "line": 0,
    "offset": 0,
  },
}
`;
